### PR TITLE
Unrendered sections does not throw when redefined and rendered in nested

### LINKED ISSUE #2252: Unrendered sections does not throw when redefined and rendered in nested layout.
Follow up #2226 
For sections that are redefined but not rendered in the outer definition.
```C#
// NestedLayout
@{ Layout = "OuterLayout"; }
@section scripts {
   // stuff
}

// OuterLayout
@{ Layout = "BaseLayout"; }
@section scripts { 
  // Don't render inner scripts
}

// BaseLayout
@RenderSection("scripts");
```
In this scenario, the scripts section defined by NestedLayout is never rendered but the view would happily succeed.

cc' @pranavkm

### PR DESCRIPTION
**Pull request by pranavkm**
layout.

Fixes #2252

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @ajaybhargavb

**Comment by ajaybhargavb** [comment]
Not fully related to this PR: Just checked MVC 5 supports having `RenderBody` inside a section which is rendered in another layout. Do we plan to have the same support? Right now our code will throw when there is no explicit `RenderBody` call.

**Comment by pranavkm** [comment]
Can't think of any reason why this might be problematic. @Eilon, any thoughts on @ajaybhargavb's scenarios?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 1]
```diff
@@ -735,6 +736,22 @@ public void EnsureBodyWasRendered()
...
         }
 
+        /// <inheritdoc />
+        public void EnsureSectionsWereRendered()
+        {
+            // If PreviousSectionWriters is set, ensure all defined sections were rendered.
+            if (PreviousSectionWriters != null)
+            {
+                var sectionsNotRendered = PreviousSectionWriters.Keys.Except(_renderedSections,
+                                                                             StringComparer.OrdinalIgnoreCase);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: nit: stylish, yes; maintainable indentation, no

**Comment by Eilon** [comment]
@ajaybhargavb so you're saying in MVC 5 a "random" section could call RenderBody() and that would work? If that's the case I'm guessing that was quite unintentional... If there's a breaking change in that scenario I'm totally fine shipping with that for now and if it turns out someone cared we can revisit... but I'd be surprised.

**Comment by ajaybhargavb** [comment]
Actually it doesn't throw even when there is no `RenderBody()` call. It is ignored. It throws only for unrendered sections. @pranavkm, was this behavior change intended in MVC 6?

**Comment by Eilon** [comment]
I *think* RenderBody was meant to be optional because there might not be an actual body, so requiring rendering it would be bad. Some users always use explicit sections, whereas some use body+sections, and some use body only.

**Comment by pranavkm** [comment]
Does seem like a change in behavior - we only required bodies to be rendered if no sections were defined. I can address this as part of the change.

**Comment by ajaybhargavb** [comment]
:+1:

**Comment by Eilon** [comment]
Ohh yeah that's right there was that logic.

**Comment by pranavkm** [comment]
Updated.

**Comment by ajaybhargavb** [comment]
:shipit:

**Comment by pranavkm** [comment]
cc @dougbu

**Comment by dougbu** [comment]
:shipit: (didn't ask for you to wait for me)

**Pull request closed without merging**